src/main/java/io/aiven/kafka/connect/common/output/OutputWriter.java

```diff
@@ -18,40 +18,47 @@
 
 import java.io.IOException;
 import java.io.OutputStream;
+import java.util.Collection;
 import java.util.Objects;
 
 import org.apache.kafka.connect.sink.SinkRecord;
 
-public abstract class OutputWriter {
+import io.aiven.kafka.connect.common.config.OutputField;
 
-    protected OutputStream outputStream;
-    protected Boolean isOutputEmpty;
+public abstract class OutputWriter implements AutoCloseable {
 
-    public OutputWriter(final OutputStream outputStream) {
+    private OutputStream outputStream;
+    private Boolean isOutputEmpty;
+    private Boolean isClosed;
+    private OutputStreamWriter writer;
+
+    public OutputWriter(final Collection<OutputField> fields,
+                        final OutputStream outputStream) {
+        this.writer = writer(fields);
         this.outputStream = outputStream;
         this.isOutputEmpty = true;
+        this.isClosed = false;
     }
 
     public void writeRecord(final SinkRecord record) throws IOException {
         Objects.requireNonNull(record, "record cannot be null");
         if (!this.isOutputEmpty) {
-            writeRecordsSeparator();
+            writer.writeRecordsSeparator(outputStream);
         } else {
-            startWriting();
+            writer.startWriting(outputStream);
             this.isOutputEmpty = false;
         }
-        writeOneRecord(record);
-    }
-
-
-    protected void startWriting() throws IOException {
+        writer.writeOneRecord(outputStream, record);
     }
 
-    protected void writeRecordsSeparator() throws IOException {
-    }
-
-    protected void writeOneRecord(final SinkRecord record) throws IOException {
-
+    public void close()  throws IOException {
+        if (!isClosed) {
+            writer.stopWriting(outputStream);
+            this.outputStream.flush();
+            this.outputStream.close();
```

Comment:
best practice close any stream using try-finally block. the reason is that java uses OS native calls for writing/reading operations. So in case of any problems with flush method you will get dangling file ids on OS level.